distribution/lib/Standard/Google_Api/0.0.0-dev/src/Analytics/Google_Analytics.enso

```diff
@@ -0,0 +1,83 @@
+polyglot java import com.google.auth.oauth2.GoogleCredentials
+polyglot java import com.google.api.gax.core.FixedCredentialsProvider
+polyglot java import com.google.api.gax.core.CredentialsProvider
+
+
+## PRIVATE
+make_dimensions_vector_selector : Widget
+make_dimensions_vector_selector =
+    item_editor = make_single_choice ['country', 'date', 'isConversionEvent', 'language', 'userAgeBracket', 'userGender']
+    Vector_Editor item_editor=item_editor item_default=item_editor.values.first.value display=Display.Always
+
+## PRIVATE
+make_metrics_vector_selector : Widget
+make_metrics_vector_selector =
+    item_editor = make_single_choice ['activeUsers', 'bounceRate', 'conversions', 'newUsers', 'sessionsPerUser', 'userConversionRate']
+    Vector_Editor item_editor=item_editor item_default=item_editor.values.first.value display=Display.Always
+
+## GROUP Standard.Base.Input
+   ICON data_input
+   Performs google analytics call
+   This method calls the google reporting v4 api. 
+   Arguments:
+   - property_id: The Google Analytics property_id that is being queried
+   - start_date: The beginning date of the query. Default is 2020-03-31
+   - end_date: The end date being queried. Defaults to today.
+@dimensions make_dimensions_vector_selector
+@metrics make_metrics_vector_selector
+read : Text -> (Vector Text) -> (Vector Text) -> Date -> Date -> Google_Credential -> Table
+read property_id:Text dimensions:Vector=['country'] metrics:Vector=['activeUsers'] start_date:Date=(Date.new 2020 3 31) end_date:Date=Date.today credentials=Google_Credential.Default -> Table =
+    analytics_data = case credentials of 
+        Google_Credential.Default -> BetaAnalyticsDataClient.create
+        Google_Credential.From_File _ -> 
+            betaAnalyticsDataSettings = BetaAnalyticsDataSettings.newBuilder
+               . setCredentialsProvider credentials.as_java
+               . build
+            BetaAnalyticsDataClient.create betaAnalyticsDataSettings
+    request_builder = RunReportRequest.newBuilder
+        . setProperty ("properties/"+property_id)
+        . addDateRanges (DateRange.newBuilder.setStartDate start_date.to_text . setEndDate end_date.to_text)
+    dimensions.distinct.each dimension->
+        request_builder.addDimensions (Dimension.newBuilder.setName dimension)
+    metrics.distinct.each metric->
+        request_builder.addMetrics (Metric.newBuilder.setName metric)
+
+    request = request_builder.build
+    response = analytics_data.runReport request
+
+    dimension_count = response.getDimensionHeadersCount
+    dimension_headers = 0.up_to dimension_count . map i-> response.getDimensionHeaders i . getName
+    metric_count = response.getMetricHeadersCount
+    metric_headers = 0.up_to metric_count . map i-> response.getMetricHeaders i . getName
+    headers = dimension_headers + metric_headers
+
+    row_count = response.getRowCount
+    row_proxy = Array_Proxy.new row_count i->
```

Comment:
What do you mean by "lookup"? Calling `.at`? I see [no caching](https://github.com/enso-org/enso/blob/9d988e90da5a4cc64eccaa7c36a5d7d62a07a31f/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayProxy.java#L70) at all. CCing @radeusgd

Reply:
There is no caching, which annotations suggest there might be?

Reply:
Anything I need to do here?

Reply:
My first thought is - No, I don't think so, it seems alright. But by the end of this I change my mind 😅 

Just to add some context.

Indeed `Array_Proxy` does no caching, it re-computes the value upon _every_ access and that is **by design**.

The intention of `Array_Proxy` (as it's name may suggest) is being able to adapt some other existing structure to 'look like an array'. So we usually put very cheap operations for the `at` method, like `i-> my_java_result.getValue i`. In these use-cases there is no sense in caching, as the operation itself is very cheap and any kind of cache would just duplicate the used memory for no apparent benefit.

We could create something like `Lazy_Vector` if we want the behaviour of computing entries lazily and caching them.

Here indeed we use the `Array_Proxy` in a slightly non-standard way because the operation is relatively expensive, as it includes `O(dimension_count + metric_count)` loops in it.

The problem is with `Table.from_array`. I thought that it accesses each row exactly once and then there really is no worry about it being proxied or not.

BUT looking at the code
```ruby
    from_rows : Vector -> Vector -> Table
    from_rows header rows =
        columns = header.map_with_index i-> name-> [name, rows.map (_.at i)]
        Table.new columns
```

We can see that what happens is we actually access each row `header.length` times.

Let `N = dimension_count + metric_count = headers.length` and `M = row_count`. Then this code actually unnecessarily becomes `O(M * N^2)` - because for every column we iterate through all rows to get the ith entry of the vector generated by the array proxy. But this runs the whole array proxy code for that row - generating the whole vector of length N just to give us its ith element.

So indeed, the `Array_Proxy` usage here seems wrong.

And the solution is very simple too - if instead of `Array_Proxy.new` we use `Vector.new` (which has exactly the same signature), it will compute every row _once_ and the whole algorithm will become `O(M * N)` which is optimal for creating an `N x M` table.

The take-away is - let's not use `Array_Proxy` for costly computations, as that is risky. Using a lazy vector here would also be useless because _every_ cell is computed anyway. I think `Vector.new` is the only right choice in this scenario. Or we'd have to rewrite `Table.from_rows` to access each row only once, but that seems like an odd requirement.

Reply:
Nice 👍 Fixed